### DISCUSSION TITLE
Inversion of background and object in label

### DISCUSSION THREAD

**Discussion by Yangtting** [Project-MONAI/tutorials#669] [Q&A] [answered]
Hello,  I use code below to save predicted label into nii.gz.
When I open the saved “nii.gz" file, I find that the background and objects in the label are completely reversed, 
"   post_transforms = Compose([
        EnsureTyped(keys="pred"),
        Invertd(
            keys="pred",
            transform=test_org_transforms,
            orig_keys="image",
            meta_keys="pred_meta_dict",
            orig_meta_keys="image_meta_dict",
            meta_key_postfix="meta_dict",
            nearest_interp=False,
            to_tensor=True,
        ),
        AsDiscreted(keys="pred", argmax=True, to_onehot=2), 
        SaveImaged(keys="pred", meta_keys="pred_meta_dict", output_dir="./out", output_postfix="seg", resample=False), ])
      
        test_data = [post_transforms(i) for i in decollate_batch(test_data)]
"
But I visualize the predicted results using code below, it shows correct. 
"            test_output = from_engine(["pred"])(test_data)
    
            original_image = loader(test_data[0]["image_meta_dict"]["filename_or_obj"])[0]
    
             plt.figure("check", (18, 6))
             plt.subplot(1, 2, 1)
             plt.imshow(original_image[:, :, 168], cmap="gray")
             plt.subplot(1, 2, 2)
             plt.imshow(test_output[0].detach().cpu()[1, :, :, 168])
             plt.show()
"
Also, when I import the saved "nii.gz" file and the original image into ITKSNAP, an error occurs, but I checked that their sizes match.
Could anyone help me?

**Comment by Nic-Ma** [comment]
Hi @Yangtting ,

Could you please help provide the MONAI version you are using?

Hi @dongyang0122 @ericspod ,

I remember we solved some similar issues before, do you remember more details?
Maybe something related to the image orientation definition.

Thanks in advance.

**Comment by dongyang0122** [comment]
It might be related to the way how pre-transforms are defined. @Yangtting It would be better if you can share your full transform list.

And the following issue may help here.
https://github.com/Project-MONAI/tutorials/issues/545

**Reply by Yangtting** [reply]
Here is my code, my MONAI version: 0.8.0, Thank you very much!
test_images = sorted(
        glob.glob(os.path.join(data_dir, "images", "*.nii.gz")))

    test_data = [{"image": image} for image in test_images]
    
    
    test_org_transforms = Compose(
        [
            LoadImaged(keys="image"),
            EnsureChannelFirstd(keys="image"),
            Orientationd(keys=["image"], axcodes="RAS"),
            ScaleIntensityRanged(
                keys=["image"], a_min=100, a_max=800,
                b_min=0.0, b_max=1.0, clip=True,
            ),
            CropForegroundd(keys=["image"], source_key="image"),
            EnsureTyped(keys="image"),
        ]
    )
    
    test_org_ds = Dataset(
        data=test_data, transform=test_org_transforms)
    
    test_org_loader = DataLoader(test_org_ds, batch_size=1, num_workers=0)
    
    post_transforms = Compose([
        EnsureTyped(keys="pred"),
        Invertd(
            keys="pred",
            transform=test_org_transforms,
            orig_keys="image",
            meta_keys="pred_meta_dict",
            orig_meta_keys="image_meta_dict",
            meta_key_postfix="meta_dict",
            nearest_interp=False,
            to_tensor=True,
        ),
        AsDiscreted(keys="pred", argmax=True, to_onehot=3), 
        SaveImaged(keys="pred", meta_keys="pred_meta_dict", output_dir="./out", output_postfix="seg", resample=False),
    ])
    # # uncomment the following lines to visualize the predicted results
    from monai.transforms import LoadImage
    loader = LoadImage()
    device = torch.device("cuda:0")
    model = UNet(
        dimensions=3,
        in_channels=1,
        out_channels=3,
        channels=(16, 32, 64, 128, 256),
        strides=(2, 2, 2, 2),
        num_res_units=2,
        norm=Norm.BATCH,
    ).to(device)
    os.chdir(savepath)

    model.load_state_dict(torch.load(modelpath))
    model.eval()
    
    with torch.no_grad():
        for test_data in test_org_loader:
            test_inputs = test_data["image"].to(device)
            roi_size = (160, 160, 160)
            sw_batch_size = 4
            test_data["pred"] = sliding_window_inference(
                test_inputs, roi_size, sw_batch_size, model)
    
            test_data = [post_transforms(i) for i in decollate_batch(test_data)]

**Comment by dongyang0122** [comment, marked as answer]
@Yangtting the pre-transforms looks good. But I notice that you have set  "to_onehot" to 3 in the post-transform. You can try to set it as None to see if it works. When "to_onehot" is not None, it will create additional dimension for classes.

https://docs.monai.io/en/stable/transforms.html#monai.transforms.AsDiscreted

**Reply by Yangtting** [reply]
Thank you for your reply, when I set ”to_onehot“ to None, my bug was solved!

When I trained the model before, I set it to 3(two targets and one background), it works well. So why I coudn't set it to 3 in testing process? Could you please explain it for me? Thanks a lot.

**Comment by dongyang0122** [comment]
For model training, loss functions require one-hot ground truth in general. But when saving the prediction, one-hot predictions are not necessary for visualization as the one-hot operation would create a 4D matrix instead of a 3D one.

**Reply by Yangtting** [reply]
Thank you! You are really nice!

**Reply by rafiibnsultan** [reply]
Thank you! Was searching for this for 2 days
